funsor/distributions.py

```diff
@@ -88,69 +91,111 @@ def eager_reduce(self, op, reduced_vars):
-        inputs, tensors = align_tensors(*params.values())
-        params = dict(zip(params, tensors))
+    def eager_log_prob(cls, *params):
+        inputs, tensors = align_tensors(*params)
+        params = dict(zip(cls._ast_fields, tensors))
         value = params.pop('value')
         data = cls.dist_class(**params).log_prob(value)
         return Tensor(data, inputs)
 
+    def __getattribute__(self, attr):
+        if attr in type(self)._ast_fields and attr != 'name':
+            return self.params[attr]
+        return super().__getattribute__(attr)
+
+    @classmethod
+    @functools.lru_cache(maxsize=None)
+    def _infer_value_domain(cls, **kwargs):
+        # rely on the underlying distribution's logic to infer the event_shape given param domains
+        instance = cls.dist_class(**{k: _dummy_tensor(domain) for k, domain in kwargs.items()}, validate_args=False)
+        out_shape = instance.event_shape
+        if isinstance(instance.support, constraints._IntegerInterval):
+            out_dtype = int(instance.support.upper_bound + 1)
+        else:
+            out_dtype = 'real'
+        return Domain(dtype=out_dtype, shape=out_shape)
+
 
 ################################################################################
 # Distribution Wrappers
 ################################################################################
 
-class BernoulliProbs(Distribution):
-    """
-    Wraps :class:`pyro.distributions.Bernoulli` .
+def make_dist(pyro_dist_class, param_names=()):
 
-    :param Funsor probs: Probability of 1.
-    :param Funsor value: Optional observation in ``{0,1}``.
-    """
-    dist_class = dist.Bernoulli
+    if not param_names:
+        param_names = tuple(name for name in inspect.getfullargspec(pyro_dist_class.__init__)[0][1:]
+                            if name in pyro_dist_class.arg_constraints)
 
-    @staticmethod
-    def _fill_defaults(probs, value='value'):
-        probs = to_funsor(probs)
-        assert probs.dtype == "real"
-        value = to_funsor(value, reals())
-        return probs, value
+    @makefun.with_signature(f"__init__(self, {', '.join(param_names)}, value='value')")
+    def dist_init(self, **kwargs):
+        return Distribution.__init__(self, *tuple(kwargs[k] for k in self._ast_fields))
 
-    def __init__(self, probs, value=None):
-        super(BernoulliProbs, self).__init__(probs, value)
+    dist_class = DistributionMeta(pyro_dist_class.__name__.split("__")[-1], (Distribution,), {
+        'dist_class': pyro_dist_class,
+        '__init__': dist_init,
+    })
 
+    eager.register(dist_class, *((Tensor,) * (len(param_names) + 1)))(dist_class.eager_log_prob)
 
-@eager.register(BernoulliProbs, Tensor, Tensor)
-def eager_bernoulli(probs, value):
-    return BernoulliProbs.eager_log_prob(probs=probs, value=value)
+    return dist_class
 
 
-class BernoulliLogits(Distribution):
-    """
-    Wraps :class:`pyro.distributions.Bernoulli` .
+class __BernoulliProbs(dist.Bernoulli):
```

Comment:
Is there a reason you're using double underscore here? I try to avoid double underscore because it makes debugging difficult due to Python's name mangling (double underscores are generally avoided for that reason).

Reply:
I changed it to avoid confusion in the wrapping logic with any custom distribution types (like Pyro's `_Subsample`) that have a `_` prefix for some reason, but I'll change it to something other than double underscore.